src/main/java/tech/ydb/io/r2dbc/executor/QueryExecutor.java

```diff
@@ -0,0 +1,261 @@
+                                    insideTransactionState.getId(),
+                                    withStatementTimeout(new CommitTxSettings())))
+                            .flatMap(ResultExtractor::extract)
+                            .doOnSuccess(unused -> updateState(connectionState.withCommitTransaction())))
+                    .then();
+        });
+    }
+
+    public Mono<Void> rollbackTransaction() {
+        return monoWithState(connectionState -> {
+            if (!connectionState.isInTransaction()) {
+                return Mono.empty();
+            }
+            final InsideTransactionState insideTransactionState = (InsideTransactionState) connectionState;
+
+            return connectionState.getSession()
+                    .flatMap(session -> Mono.fromFuture(session.rollbackTransaction(
+                                    insideTransactionState.getId(),
+                                    withStatementTimeout(new RollbackTxSettings())))
+                            .flatMap(ResultExtractor::extract)
+                            .doOnSuccess(unused -> updateState(connectionState.withRollbackTransaction())))
+                    .then();
+        });
+    }
+
+    public Mono<Void> updateState(Function<YdbConnectionState, YdbConnectionState> function) {
+        return Mono.fromRunnable(() -> this.ydbConnectionState = function.apply(this.getCurrentState()));
+    }
+
+    public YdbConnectionState getCurrentState() {
+        return ydbConnectionState;
+    }
+
+    public Mono<Void> setAutoCommit(boolean autoCommit) {
+        return monoWithState(connectionState -> {
+
+            if (connectionState.getYdbTxSettings().isAutoCommit() == autoCommit) {
+                return Mono.empty();
+            }
+
+            if (autoCommit && connectionState.isInTransaction()) {
+                return commitTransaction()
+                        .doOnSuccess((unused) -> updateState(connectionState.withAutoCommit(true)));
+            }
+
+            updateState(connectionState.withAutoCommit(autoCommit));
+
+            return Mono.empty();
+        });
+    }
+
+    public Mono<Void> close() {
+        return monoWithState(connectionState -> {
+            if (connectionState.isInTransaction()) {
+                return commitTransaction()
+                        .doOnSuccess(unused -> updateState(connectionState.close()));
+            }
+
+            updateState(connectionState.close());
+
+            return Mono.empty();
+        });
+    }
+
+    public Mono<Void> setStatementTimeout(Duration timeout) {
+        return Mono.fromRunnable(() -> this.statementTimeout = timeout);
+    }
+
+    private void updateState(YdbConnectionState ydbConnectionState) {
+        this.ydbConnectionState = ydbConnectionState;
+    }
+
+    private <T> Mono<T> monoWithState(Function<YdbConnectionState, Mono<T>> function) {
+        final YdbConnectionState currentYdbConnectionState = ydbConnectionState;
```

Comment:
Clarify what happens before in the r2dbc specification

async call is await or changeset

Reply:
Monitoring the state of the object "The connection object is unsafe for simultaneous state changes by several subscribers. The connection object can be shared by multiple threads that perform operations sequentially using the appropriate synchronization mechanisms." That is, there is no guarantee that the next method will be called after Publisher closes the previous method. Such a guarantee is too strict, for example, such a guarantee will not allow running multiple database queries at the same time that are not connected to each other in the same connection.
```
Statement statement1 = connection.createStatement("select * from people;");
Statement statement2 = connection.createStatement("select * from orders;");

 Flux.zip(statement1.execute(), statement2.execute());
```
The scenario when a request is executed in one connection before the full start of the transaction seems strange and incorrect.
```
Statement statement = connection.createStatement("select * from orders;");

Flux.zip(connection.beginTransaction(), statement.execute());
```
I think we should provide a defined behavior for which we fix the state and work with it. The same is done in r2dbc-postgresql. https://github.com/pgjdbc/r2dbc-postgresql/blob/main/src/main/java/io/r2dbc/postgresql/Postgresql Connection.java#L451